### ISSUE TITLE
cpu.write_mem to MMIO devices is ignored

### ISSUE THREAD

**Issue by vringar** [AFLplusplus/qemu-libafl-bridge#95]
As write_mem used the `cpu_memory_rw_debug` function it can not write to MMIO devices. [The relevant upstream issue](https://gitlab.com/qemu-project/qemu/-/issues/213)
I needed this for my fuzzing, so I adapted the proposed patch and it works for me.
I'm unable to engage with the QEMU mailing list, but I'd happily upstream this to you, if you are interested in taking it.

**Comment by rmalmain** [comment]
indeed, thanks for the proposition. i think it would be nice to expose this in some way.
do you mind opening a pr with the changes you did? i can have a look at your patch.

**Comment by vringar** [comment]
I've challenged myself and am currently trying to get the patch directly into QEMU as well [v1](https://lists.nongnu.org/archive/html/qemu-devel/2024-12/msg04622.html) [v2](https://lists.nongnu.org/archive/html/qemu-devel/2024-12/msg05258.html)

**Comment by rmalmain** [comment]
good, it would be better to upstream this to QEMU directly.
in that case, i think it's better to see how it goes on their side before merging your PR.
feel free to post updates there, whether it gets merged to mainline or not

**Comment by rmalmain** [comment]
hey @vringar, do you have any news regarding the patch?
if not, i think we can merge it there while they decide to merge it or not in mainline.

**Comment by vringar** [comment]
[The patch set](https://lists.nongnu.org/archive/html/qemu-devel/2025-02/msg01746.html) triggered by this request has been merged into a subtree. I couldn't figure out how to trace it from there to mainline but I believe it should be on it's way

**Comment by rmalmain** [comment]
great, thanks for the update.
i'll close this issue and the PR once we merge back the changes with the next qemu release.
